development/extensions/tutorial_php_events_and_extending_extensions.rst

```diff
@@ -0,0 +1,305 @@
+registered as services.
+
+The ``\phpbb\extension\finder`` is available from the service container as
+``ext.finder`` and can be used as follows. The following example is part of what
+is used to find all routing files, both for extensions and core routes.
+
+.. code-block:: php
+
+  $finder
+      ->directory('config')
+      ->suffix('routing.yml')
+      ->find();
+
+As you can see, you are able to chain method calls together (ending with ``find()``).
+Check out the class definition for more information about the different methods
+that are available (such as specifying a different directory for extensions than
+core files, and getting class names based on files it finds). The return of the
+``find()`` method is an array of file paths that match the given criteria.
+
+For example, assume you have an extension and you need to get a collection of all
+PNG image files from one of its directories. You could use the extension manager
+to load the finder, and traverse the extension's image directory as follows:
+
+.. code-block:: php
+
+  $finder = $extension_manager->get_finder();
+
+  $images = $finder
+      ->extension_suffix('.png')
+      ->extension_directory('/images')
+      ->find_from_extension('demo', $phpbb_root_path . 'ext/acme/demo/');
+
+The ``$images`` array would look something like (as you can see the image paths are
+contained in the array keys):
+
+.. code-block:: php
+
+  array(
+      'ext/acme/demo/images/image1.png' => 'demo',
+      'ext/acme/demo/images/image2.png' => 'demo',
+  );
+
+
+Using service replacement
+=========================
+.. warning::
+  Only use service replacements if your extensions needs to be compatible with
+  phpBB 3.1. For 3.2 and later, see `Using service decoration`_.
+
+.. warning::
+  You can't assume the order of a service is defined in phpBB if other extensions
+  are installed. Be really carefully when using service replacements.
+
+With a service replacement you can replace a existing service in phpBB (Or in an
+extension) with your own service. Your replacement should type match the original
+service (so if the original service implements a interface, you should minimal
+implement that specific interface, if it is a concrete class, you will need to
+extent that class). The best way to accomplish this is by extending the original
+class, and only replace the features you want to change.
+
+To replace a core phpBB service, you simply name your extensions service name
+with the same name as the service in phpBB's core.
+
+If you want, for example, replace the ``config`` service in phpBB with your own
+implementation, your service configuration will look like this:
+
+.. code-block:: yaml
+
+  config:
+      class: acme\demo\config\db
+      arguments:
+          - '@dbal.conn'
+          - '@cache.driver'
+          - '%tables.config%'
+          - '@acme.demo.db_reader'
+
+Your php class should extends, at least, ``\phpbb\config\config``, which is the base
```

Comment:
Your PHP class could extend ``\phpbb\config\config`` which is the base class phpBB extends from. However, to avoid any unforseen problems, it's better to more explicitly extend from the ``\phpbb\config\db`` class, which is the specific class being replaced by the service.